src/installer.ts

```diff
@@ -30,6 +30,143 @@ interface INodeRelease extends tc.IToolRelease {
   lts?: string;
 }
 
+export enum Distributions {
+  DEFAULT,
+  CANARY,
+  NIGHTLY,
+  RC
+}
+
+export const distributionOf = (versionSpec: string): Distributions =>
+  versionSpec.includes('-v8-canary')
+    ? Distributions.CANARY
+    : // TODO: i'd like to have this check, do you?
+    versionSpec.includes('-canary')
+    ? (() => {
+        throw Error('Canary version must have "-v8-canary suffix"');
+      })()
+    : versionSpec.includes('nightly')
+    ? Distributions.NIGHTLY
+    : semver.prerelease(versionSpec)
+    ? Distributions.RC
+    : Distributions.DEFAULT;
+
+interface VersionMatcher {
+  (potential: string): boolean;
+
+  // memoize the factory for testing and debug purposes
+  factory:
+    | ((ver: string, suffix: string) => VersionMatcher)
+    | ((semverRanger: string) => VersionMatcher)
+    | (() => VersionMatcher);
+}
+
+export const semverVersionMatcherFactory = (range: string): VersionMatcher => {
+  const matcher = (potential: string): boolean =>
+    semver.satisfies(potential, range);
+  matcher.factory = semverVersionMatcherFactory;
+  return matcher;
+};
+
+export const canaryRangeVersionMatcherFactory = (
+  version: string
+): VersionMatcher => {
+  const range = semver.validRange(`^${version}`);
+  const matcher = (potential: string): boolean =>
+    semver.satisfies(potential.replace('-v8-canary', '+v8-canary.'), range);
+  matcher.factory = canaryRangeVersionMatcherFactory;
+  return matcher;
+};
+
+export const canaryExactVersionMatcherFactory = (
+  version: string,
+  timestamp: string
+): VersionMatcher => {
+  const range = `${version}-${timestamp}`;
+  const matcher = (potential: string): boolean =>
+    semver.satisfies(potential, range);
+  matcher.factory = canaryExactVersionMatcherFactory;
+  return matcher;
+};
+
+export const nightlyRangeVersionMatcherFactory = (
+  version: string
+): VersionMatcher => {
+  const range = semver.validRange(`^${version}`);
+  // TODO: this makes v20.1.1-nightly to do not match v20.1.1-nightly20221103f7e2421e91
+  // const range = `${semver.validRange(`^${version}-0`)}-0`;
+  const matcher = (potential: string): boolean =>
+    distributionOf(potential) === Distributions.NIGHTLY &&
+    // TODO: dmitry's variant was potential.replace('-nightly', '-nightly.') that made
+    //       all unit tests to fail
```

Comment:
Could you please share an example of their fallings to investigate it deeper because I'd like to not split function for raw and range.

Reply:
```
const potential='v20.0.0-nightly20221103f7e2421e91'
const raw='v20'
const validVersion = semver.valid(raw) ? raw : semver.coerce(raw)?.version;
const range = semver.validRange(`^${version}`);
semver.satisfies(potential.replace('-nightly', '-nightly.'),range) // false
semver.satisfies(potential.replace('-nightly', '-nightly.'),range,{includePrerelease: true}) // false

const range0 = `${semver.validRange(`^${version}-0`)}-0`;
semver.satisfies(potential.replace('-nightly', '-nightly.'),range0) // false
semver.satisfies(potential.replace('-nightly', '-nightly.'),range0,{includePrerelease: true}) // false

semver.satisfies(potential.replace('-nightly', '+nightly.'),range) // true
```